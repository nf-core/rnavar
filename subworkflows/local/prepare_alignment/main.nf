//
// Prepare input alignment files
//

include { SAMTOOLS_CONVERT } from '../../../modules/nf-core/samtools/convert'
include { SAMTOOLS_INDEX   } from '../../../modules/nf-core/samtools/index'

workflow PREPARE_ALIGNMENT {
    take:
    cram     // [ val(meta), path(cram), path(crai) ]
    bam      // [ val(meta), path(bam), path(bai) ]

    main:
    ch_versions = Channel.empty()

    def alignment_branch = bam
        .mix(cram)
        .branch { meta, bam_, bai ->
            indexed: bai
                return [ meta, bam_, bai ]
            not_indexed_bam: !bai && bam_.extension == "bam"
                return [ meta, bam_ ]
            not_indexed_cram: !bai && bam_.extension == "cram"
                return [ meta, bam_ ]
        }

    def bam_no_index = alignment_branch.not_indexed_bam.mix(alignment_branch.not_indexed_cram)

    SAMTOOLS_INDEX(
        bam_no_index
    )
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

    def bam_indexed = alignment_branch.not_indexed_bam
        .join(SAMTOOLS_INDEX.out.bai, failOnMismatch: true, failOnDuplicate: true)

    def cram_indexed = alignment_branch.not_indexed_cram
        .join(SAMTOOLS_INDEX.out.crai, failOnMismatch: true, failOnDuplicate: true)

    def alignment_out = bam_indexed
        .mix(cram_indexed)
        .mix(alignment_branch.indexed)

    emit:
    bam      = alignment_out // [ val(meta), path(bam), path(bai) ]
    versions = ch_versions
}
