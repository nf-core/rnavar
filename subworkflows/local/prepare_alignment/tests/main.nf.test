nextflow_workflow {

    name "Test Subworkflow PREPARE_ALIGNMENT"
    script "../main.nf"
    workflow "PREPARE_ALIGNMENT"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/prepare_alignment"
    tag "samtools/convert"
    tag "samtools/index"

    test("only bam") {
        when {

            workflow {
                """
                input[0] = Channel.of([
                    [id:'test'],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.collectEntries { key, val ->
                        if(key.matches("\\d+")) {
                            return null
                        } else if(key == "bam") {
                            return [key, val.collect { [it[0], it[1], file(it[2]).name]}]
                        }
                        return [key,val]
                    }.findAll { it != null }
                ).match() }
            )
        }
    }

    test("only cram - no index") {
        when {

            workflow {
                """
                input[0] = Channel.empty()
                input[1] = Channel.of([
                    [id:'test'],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.rna.paired_end.sorted.cram', checkIfExists: true),
                    []
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.collectEntries { key, val ->
                        if(key.matches("\\d+")) {
                            return null
                        } else if(key == "bam") {
                            return [key, val.collect { [it[0], it[1], file(it[2]).name]}]
                        }
                        return [key,val]
                    }.findAll { it != null }
                ).match() }
            )
        }
    }

    test("bam, cram - no index, index") {
        when {

            workflow {
                """
                input[0] = Channel.of([
                    [id:'test'],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam', checkIfExists: true),
                    []
                ])
                input[1] = Channel.of([
                    [id:'test'],
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.rna.paired_end.sorted.cram', checkIfExists: true),
                    file(params.pipelines_testdata_base_path + '/genomics/homo_sapiens/illumina/cram/test.rna.paired_end.sorted.cram.crai', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(
                    workflow.out.collectEntries { key, val ->
                        if(key.matches("\\d+")) {
                            return null
                        } else if(key == "bam") {
                            return [key, val.collect { [it[0], it[1], file(it[2]).name]}]
                        }
                        return [key,val]
                    }.findAll { it != null }
                ).match() }
            )
        }
    }
}
