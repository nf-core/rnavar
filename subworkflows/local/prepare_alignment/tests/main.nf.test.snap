{
    "only cram - no index": {
        "content": [
            {
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test.rna.paired_end.sorted.cram",
                        "test.rna.paired_end.sorted.cram.crai"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b9253c084ace536ae84f4df30fe71912"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-05T13:36:56.176207564"
    },
    "bam, cram - no index, index": {
        "content": [
            {
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam",
                        "test.rna.paired_end.sorted.bam.bai"
                    ],
                    [
                        {
                            "id": "test"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/cram/test.rna.paired_end.sorted.cram",
                        "test.rna.paired_end.sorted.cram.crai"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,b9253c084ace536ae84f4df30fe71912"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-05T13:37:07.592019675"
    },
    "only bam": {
        "content": [
            {
                "bam": [
                    [
                        {
                            "id": "test"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.rna.paired_end.sorted.bam",
                        "test.rna.paired_end.sorted.bam.bai"
                    ]
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-03-05T13:32:21.806173639"
    }
}